File: strategy/typescript/glados/test/unit/base/processor.lua
local Injector = require "test/unit/injector"

test("base.process", function()
	local Process = require "../base/process"
	local instance = Process()
	assert_error(function() instance:run() end, "Expected error from stub")
	assert_error(function() instance:isFinished() end, "Expected error from stub")
end)

context("base.processor", function()
	local Processor, Process, SpyProcess, Class

	before(function()
		Class = Injector.newClassLoader()
		local injector = Injector(Class, true)
		Process = injector:load("../base/process")

		SpyProcess = Class("SpyProcess", Process)
		function SpyProcess:init()
			self.counter = 0
			self.finished = false
		end
		function SpyProcess:run() self.counter = self.counter + 1 end
		function SpyProcess:isFinished() return self.finished end

		Processor = injector:load("../base/processor")
	end)

	test("invalid type", function()
		-- must be of type processor
		assert_error(function() Processor.addPre({}) end)
		assert_error(function() Processor.addPost({}) end)
	end)

	test("correct pre and post", function()
		local preInstance = SpyProcess()
		local postInstance = SpyProcess()
		Processor.addPre(preInstance)
		assert_equal(preInstance.counter, 0)
		Processor.pre()
		assert_equal(preInstance.counter, 1)
		assert_equal(postInstance.counter, 0)
		Processor.post()
		assert_equal(preInstance.counter, 1)
		assert_equal(postInstance.counter, 0)
		Processor.addPost(postInstance)
		assert_equal(preInstance.counter, 1)
		assert_equal(postInstance.counter, 0)
		Processor.pre()
		assert_equal(preInstance.counter, 2)
		assert_equal(postInstance.counter, 0)
		Processor.post()
		assert_equal(preInstance.counter, 2)
		assert_equal(postInstance.counter, 1)
	end)

	test("finished process", function()
		local instance = SpyProcess()
		Processor.addPre(instance)
		assert_equal(instance.counter, 0)
		Processor.pre()
		assert_equal(instance.counter, 1)
		instance.finished = true
		Processor.pre()
		assert_equal(instance.counter, 2)
		-- check that process is removed
		Processor.pre()
		assert_equal(instance.counter, 2)

		-- an already finished process is run exactly once
		Processor.addPre(instance)
		Processor.pre()
		assert_equal(instance.counter, 3)
		Processor.pre()
		assert_equal(instance.counter, 3)
	end)
end)

File: strategy/typescript/glados/test/unit/base/processor.ts
/**************************************************************************
*   Copyright 2026 Robotics Erlangen e.V., Tobias Heineken                *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
**************************************************************************/

import { Process } from "base/process";
import * as Processor from "base/processor";

import { UnitTest } from "glados/test/unit/unittest";

class SpyProcess implements Process {
	public counter: number = 0;
	public finished: boolean = false;

	public run() {
		this.counter += 1;
	}
	public isFinished() {
		return this.finished;
	}
}

export class BaseProcessor extends UnitTest {
	public constructor() {
		super();
		this._addTest("correct pre and post", this._testCorrectPreAndPost);
		this._addTest("finished process", this._testFinishedProcess);
	}

	private _testCorrectPreAndPost() {
		let preInstance = new SpyProcess();
		let postInstance = new SpyProcess();
		Processor.addPre(preInstance);
		this._assert_eq(preInstance.counter, 0);
		Processor.pre();
		this._assert_eq(preInstance.counter, 1);
		this._assert_eq(postInstance.counter, 0);
		Processor.post();
		this._assert_eq(preInstance.counter, 1);
		this._assert_eq(postInstance.counter, 0);
		Processor.addPost(postInstance);
		this._assert_eq(preInstance.counter, 1);
		this._assert_eq(postInstance.counter, 0);
		Processor.pre();
		this._assert_eq(preInstance.counter, 2);
		this._assert_eq(postInstance.counter, 0);
		Processor.post();
		this._assert_eq(preInstance.counter, 2);
		this._assert_eq(postInstance.counter, 1);
	}

	private _testFinishedProcess() {
		let instance = new SpyProcess();
		Processor.addPre(instance);
		this._assert_eq(instance.counter, 0);
		Processor.pre();
		this._assert_eq(instance.counter, 1);
		instance.finished = true;
		Processor.pre();
		this._assert_eq(instance.counter, 2);
		// check that process is removed
		Processor.pre();
		this._assert_eq(instance.counter, 2);

		// an already finished process is run exactly once
		Processor.addPre(instance);
		Processor.pre();
		this._assert_eq(instance.counter, 3);
		Processor.pre();
		this._assert_eq(instance.counter, 3);
	}
}
export let testClass = BaseProcessor;
